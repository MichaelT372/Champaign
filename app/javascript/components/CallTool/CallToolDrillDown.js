// @flow
import React, { Component } from 'react';
import {
  compact,
  flatten,
  get,
  omit,
  omitBy,
  sample,
  startCase,
  uniq,
} from 'lodash';
import {
  filterTargets,
  targetsWithFields,
  valuesForFilter,
} from './call_tool_helpers';
import SweetSelect from '../../components/SweetSelect/SweetSelect';
import type { TargetWithFields } from './call_tool_helpers';

type Filters = { [string]: string };
type Props = {
  targetByAttributes: string[],
  filters: Filters,
  targets: TargetWithFields[],
  onUpdate: (target: TargetWithFields) => void,
};

type State = {
  filters: Filters,
  targets: TargetWithFields[],
};
export default class CallToolDrillDown extends Component {
  props: Props;
  state: State;

  constructor(props: Props) {
    super(props);

    const filters = Object.assign({}, props.filters);
    this.state = {
      filters,
      targets: filterTargets(props.targets, filters),
    };
  }

  componentDidMount() {
    if (!this.props.targetByAttributes.length) {
      this.props.onUpdate(this.sampleTarget(this.props.targets));
    }
  }

  sampleTarget(targets: TargetWithFields[]): TargetWithFields {
    return sample(targets);
  }

  valuesForSelect(key: string): any[] {
    const values = valuesForFilter(
      this.props.targets,
      this.props.targetByAttributes,
      this.state.filters,
      key
    );
    return values.map(value => ({
      id: `${key}-${value}`,
      label: value,
      value,
    }));
  }

  updateFilters(filters: Filters) {
    this.setState(
      (prevState: State) => ({
        ...prevState,
        filters,
        targets: filterTargets(this.props.targets, filters),
      }),
      () => this.props.onUpdate(this.sampleTarget(this.state.targets))
    );
  }

  renderFilter = (key: string, index: number) => {
    const attrs = this.props.targetByAttributes;
    const previousKey = attrs[index - 1];
    const previousFilter = this.state.filters[previousKey];
    // if we're in the first filter, or if we've selected a previous filter
    if (index === 0 || previousFilter) {
      return (
        <div key={key} style={{ marginBottom: '10px' }}>
          <SweetSelect
            name={`filter-targets-by-attribute-${key}`}
            value={this.state.filters[key]}
            options={this.valuesForSelect(key)}
            label={startCase(key)}
            clearable={index > 0}
            onChange={(value: string) => {
              const updatedFilters = { ...this.state.filters, [key]: value };
              const discardFn = (v: string, k: string) =>
                !v || attrs.indexOf(k) > index;
              this.updateFilters(omitBy(updatedFilters, discardFn));
            }}
          />
          <div className="clearfix" />
        </div>
      );
    }

    return null;
  };

  render() {
    if (!this.props.targetByAttributes.length) return null;
    return (
      <div className="CallToolDrillDown">
        {this.props.targetByAttributes.map(this.renderFilter)}
      </div>
    );
  }
}
