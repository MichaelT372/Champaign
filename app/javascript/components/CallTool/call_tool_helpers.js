// @flow
import { compact, isMatch, pick, uniq } from 'lodash';
import type { Target } from '../../call_tool/CallToolView';

export function valuesForFilter(
  targets: Target[],
  attrs: string[],
  filters: { [string]: string },
  currentFilter: string
): string[] {
  const index = attrs.indexOf(currentFilter);
  const activeFilters = pick(filters, attrs.slice(0, index));
  const values = targets
    .filter(t => isMatch(t, activeFilters))
    .map(t => t[currentFilter]);

  console.debug({
    index,
    attrs,
    activeFilters,
    filters,
    currentFilter,
    targets,
    filteredValues: values,
  });

  return compact(uniq(values));
}
