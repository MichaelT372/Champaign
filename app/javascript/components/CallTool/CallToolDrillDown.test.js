import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import CallToolDrillDown from './CallToolDrillDown';
import { targets } from './call_tool_helpers.test';

const onUpdate = jest.fn();
const propsWithFiltering = {
  targets,
  onUpdate,
  targetByAttributes: ['country', 'state'],
  filters: {},
};

describe('CallToolDrillDown Snapshots', () => {
  it('renders filters when there are targetByAttributes', () => {
    const wrapper = shallow(<CallToolDrillDown {...propsWithFiltering} />);
    expect(toJSON(wrapper)).toMatchSnapshot('CallToolDrillDownWithFilters');
  });
});
