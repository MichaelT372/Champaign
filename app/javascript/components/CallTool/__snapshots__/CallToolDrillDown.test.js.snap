// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallToolDrillDownWithFilters 1`] = `
<div
  className="CallToolDrillDown"
>
  <div
    style={
      Object {
        "marginBottom": "10px",
      }
    }
  >
    <SweetSelect
      clearable={false}
      label="Country"
      name="filter-targets-by-attribute-country"
      onChange={[Function]}
      options={
        Array [
          Object {
            "id": "country-United States",
            "label": "United States",
            "value": "United States",
          },
          Object {
            "id": "country-Canada",
            "label": "Canada",
            "value": "Canada",
          },
        ]
      }
    />
    <div
      className="clearfix"
    />
  </div>
</div>
`;
